### PR TITLE
add where/orderby/limit after “SHOW ALTER TABLE COLUMN“ syntax

### PR DESCRIPTION
**Pull request by marising**
#2314

Features：
1、Support WHERE/ORDER BY/LIMIT
2、Columns：TableName、CreatTime、FinishTime、State
3、Only  “And” between conditions
4、TableName and State column only support "=" operator
5、CreateTime and FinishTime column support “=”,“>=”,"<=",">","<","!=" operators
6、CreateTime and FinishTime column support Date and DateTime string, eg:"2019-12-04" or "2019-12-04 17:18:00"

TestCase:
MySQL [haibotest]> show alter table column where State='FINISHED' and CreateTime > '2019-12-03' order by FinishTime desc limit 0,2;
+-------+---------------+---------------------+---------------------+---------------+---------+---------------+---------------+---------------+----------+------+----------+---------+
| JobId | TableName | CreateTime | FinishTime | IndexName | IndexId | OriginIndexId | SchemaVersion | TransactionId | State | Msg | Progress | Timeout |
+-------+---------------+---------------------+---------------------+---------------+---------+---------------+---------------+---------------+----------+------+----------+---------+
| 11134 | test_schema_2 | 2019-12-03 19:21:42 | 2019-12-03 19:22:11 | test_schema_2 | 11135 | 11059 | 1:192010000 | 3 | FINISHED | | N/A | 86400 |
| 11096 | test_schema_3 | 2019-12-03 19:21:31 | 2019-12-03 19:21:51 | test_schema_3 | 11097 | 11018 | 1:2063361382 | 2 | FINISHED | | N/A | 86400 |
+-------+---------------+---------------------+---------------------+---------------+---------+---------------+---------------+---------------+----------+------+----------+---------+
2 rows in set (0.00 sec)

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java [hunk 1]
```diff
@@ -121,11 +221,30 @@ private void handleShowAlterTable(Analyzer analyzer) throws AnalysisException, U
...
+                sb.append((orderByElements.get(i).getIsAsc()) ? " ASC" : " DESC");
+                sb.append((i + 1 != orderByElements.size()) ? ", " : "");
+            }
+        }
 
+        if (limitElement != null && limitElement.hasLimit()) {
+            sb.append(" LIMIT ").append(limitElement.getLimit());
+            if (limitElement.hasOffset()) {
+                sb.append(" OFFSET ").append(limitElement.getOffset());
+            }
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java]: why not sb.append(limitElement.toSql())

File: fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java [hunk 2]
```diff
@@ -121,11 +221,30 @@ private void handleShowAlterTable(Analyzer analyzer) throws AnalysisException, U
...
+            sb.append(" WHERE ").append(whereClause.toSql());
+        }
+        // Order By clause
+        if (orderByElements != null) {
+            sb.append(" ORDER BY ");
+            for (int i = 0; i < orderByElements.size(); ++i) {
+                sb.append(orderByElements.get(i).getExpr().toSql());
+                sb.append((orderByElements.get(i).getIsAsc()) ? " ASC" : " DESC");
+                sb.append((i + 1 != orderByElements.size()) ? ", " : "");
+            }
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java]: sb.append(orderByElements.toSql())

File: fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java [hunk 3]
```diff
@@ -121,11 +221,30 @@ private void handleShowAlterTable(Analyzer analyzer) throws AnalysisException, U
...
+        }
 
+        if (limitElement != null && limitElement.hasLimit()) {
+            sb.append(" LIMIT ").append(limitElement.getLimit());
+            if (limitElement.hasOffset()) {
+                sb.append(" OFFSET ").append(limitElement.getOffset());
+            }
+        }
+        LOG.info("AlterSQL '{}'", sb.toString());
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java]: should remove this log

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 4]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+                                          LimitElement limitElement) throws AnalysisException {
+        Preconditions.checkNotNull(db);
+        Preconditions.checkNotNull(schemaChangeHandler);
+
+        List<List<Comparable>> schemaChangeJobInfos = schemaChangeHandler.getAlterJobInfosByDb(db);
+
+        LOG.info("Begin fetch data.Filter size " + filter.size() );
+        for(Entry<String, Expr> entry : filter.entrySet()){
+            LOG.info("key " + entry.getKey() + ", value" + entry.getValue().getChild(1));
+        }
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]: should remove debug log

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 5]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+
+        //where
+        List<List<Comparable>> jobInfos = new ArrayList<List<Comparable>>();
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]:
```suggestion
        List<List<Comparable>> jobInfos = Lists.newArrayList();
```

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 6]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+        for (List<Comparable> infoStr : schemaChangeJobInfos) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]: if filter is null, jobInfos can be assigned with schemaChangeJobInfos without iterating schemaChangeJobInfos one by one

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 7]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+                LOG.warn("SchemaChangeJobInfos.size() " + schemaChangeJobInfos.size()
+                    + " not equal TITLE_NAMES.size() " + TITLE_NAMES.size());
+                continue;
+            }
+            List<Comparable> jobInfo = new ArrayList<Comparable>();
+            boolean isNeed = true;
+            for (int i = 0; i < infoStr.size(); i++) {
+                Comparable element = infoStr.get(i);
+                isNeed = filterResult(TITLE_NAMES.get(i), element, filter);
+                LOG.info("column "+ TITLE_NAMES.get(i) +", value " + element + ",isNeed " + isNeed);
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]: remove this log

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 8]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+        }
+        BinaryPredicate binaryPredicate = (BinaryPredicate) subExpr;
+        if (subExpr.getChild(1) instanceof StringLiteral && binaryPredicate.getOp() == BinaryPredicate.Operator.EQ) {
+            LOG.info("SubExpr value " + ((StringLiteral) subExpr.getChild(1)).getValue() + ", real value " + element);
+            return ((StringLiteral) subExpr.getChild(1)).getValue().equals(element);
+        }
+        if (subExpr.getChild(1) instanceof DateLiteral) {
+            Long leftVal = (new DateLiteral((String) element, Type.DATETIME)).getLongValue();
+            Long rightVal = ((DateLiteral) subExpr.getChild(1)).getLongValue();
+            LOG.info("Left value " + leftVal + ", right value " + rightVal);
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]:
```suggestion
```

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 9]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+        if (filter == null) {
+            return true;
+        }
+        Expr subExpr = filter.get(columnName.toLowerCase());
+        if (subExpr == null) {
+            return true;
+        }
+        BinaryPredicate binaryPredicate = (BinaryPredicate) subExpr;
+        if (subExpr.getChild(1) instanceof StringLiteral && binaryPredicate.getOp() == BinaryPredicate.Operator.EQ) {
+            LOG.info("SubExpr value " + ((StringLiteral) subExpr.getChild(1)).getValue() + ", real value " + element);
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]:
```suggestion

```

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 10]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+            return ((StringLiteral) subExpr.getChild(1)).getValue().equals(element);
+        }
+        if (subExpr.getChild(1) instanceof DateLiteral) {
+            Long leftVal = (new DateLiteral((String) element, Type.DATETIME)).getLongValue();
+            Long rightVal = ((DateLiteral) subExpr.getChild(1)).getLongValue();
+            LOG.info("Left value " + leftVal + ", right value " + rightVal);
+            switch( binaryPredicate.getOp()) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]:
```suggestion
            switch (binaryPredicate.getOp()) {
```

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 11]
```diff
@@ -42,6 +58,110 @@ public SchemaChangeProcNode(SchemaChangeHandler schemaChangeHandler, Database db
...
+            List<Comparable> jobInfo = new ArrayList<Comparable>();
+            boolean isNeed = true;
+            for (int i = 0; i < infoStr.size(); i++) {
+                Comparable element = infoStr.get(i);
+                isNeed = filterResult(TITLE_NAMES.get(i), element, filter);
+                LOG.info("column "+ TITLE_NAMES.get(i) +", value " + element + ",isNeed " + isNeed);
+                if (!isNeed) {
+                    break;
+                }
+                jobInfo.add(element.toString());
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]: why not append infoStr to jobInfos when isNeed is true, other than putting element one by one

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/cup/sql_parser.cup [hunk 12]
```diff
@@ -1932,9 +1932,9 @@ show_param ::=
         RESULT = new ShowDeleteStmt(db);
     :}
     /* Show alter table statement: used to show process of alter table statement */
-    | KW_ALTER KW_TABLE opt_alter_type:type opt_db:db
+    | KW_ALTER KW_TABLE opt_alter_type:type opt_db:db opt_wild_where order_by_clause:orderByClause limit_clause:limitClause
```
[the remark this answers is not in the document]
Comment (marising) [reply fe/src/main/cup/sql_parser.cup]: Done

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 13]
```diff
@@ -61,4 +175,12 @@ public ProcResult fetchResult() throws AnalysisException {
         return result;
     }
 
+    public static int analyzeColumn(String columnName) throws AnalysisException {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]: Maybe getColumnIndex is more appropriate.

File: fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java [hunk 14]
```diff
@@ -61,4 +175,12 @@ public ProcResult fetchResult() throws AnalysisException {
...
+        for (String title : TITLE_NAMES) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/common/proc/SchemaChangeProcNode.java]: why not `for (int i = 0; i < TITILE_NAME.size(); ++i)`

File: fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java [hunk 15]
```diff
@@ -52,28 +58,92 @@
...
         this.type = type;
         this.dbName = dbName;
+        this.whereClause = whereClause;
+        this.orderByElements = orderByElements;
+        this.limitElement = limitElement;
+        this.filterMap = new HashMap<String, Expr>();
+    }
+
+    private boolean getPredicateValue(Expr subExpr) throws AnalysisException {
+        if (!(subExpr instanceof BinaryPredicate)) {
```
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java]: should throw excption for unsupported Expr

File: fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java [hunk 16]
```diff
@@ -52,28 +58,87 @@
...
+        this.filterMap = new HashMap<String, Expr>();
+    }
+
+    private void getPredicateValue(Expr subExpr) throws AnalysisException {
+        if (!(subExpr instanceof BinaryPredicate)) {
+            return;
+        }
+        BinaryPredicate binaryPredicate = (BinaryPredicate) subExpr;
+        if (!(subExpr.getChild(0) instanceof SlotRef)) {
+            return;
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/ShowAlterStmt.java]: throw an exception?

File: docs/documentation/cn/sql-reference/sql-statements/Data Manipulation/SHOW ALTER.md [hunk 17]
```diff
@@ -25,6 +25,7 @@ under the License.
         
     说明：
         TABLE COLUMN：展示修改列的 ALTER 任务
+                      支持语法[WHERE TableName|CreateTime|FinishTime|State] [ORDER BY] [LIMIT]
```
Comment (imay) [inline docs/documentation/cn/sql-reference/sql-statements/Data Manipulation/SHOW ALTER.md]: better to give some example below

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**